ouroboros-network/test/Test/Ouroboros/Network/Testnet.hs

```diff
@@ -27,24 +64,281 @@ tests =
+                      . fmap (\(WithName _ (WithTime t b)) -> (t, b))
+                      )
+               . Trace.toList
+               . splitWithNameTrace
+               . Trace.fromList ()
+               . fmap snd
+               . Signal.eventsToList
+               . Signal.eventsFromListUpToTime (Time (10 * 60 * 60))
+               . Trace.toList
+               . fmap (\(WithTime t (WithName name b)) -> (t, WithName name (WithTime t b)))
+               . withTimeNameTraceEvents
+                  @DiffusionTestTrace
+                  @NtNAddr
+               $ runSimTrace sim
+
+     in conjoin
+      $ verify_target_established_local
+      <$> events
+
   where
-    sim :: forall s . IOSim s Void
-    sim = diffusion_simulation (toBearerInfo defaultBearerInfo) dmnScript
+    verify_target_established_local :: Events DiffusionTestTrace -> Property
+    verify_target_established_local events =
+      let govLocalRootPeersSig :: Signal (Set NtNAddr)
+          govLocalRootPeersSig =
+            selectDiffusionPeerSelectionState
+              ( LocalRootPeers.keysSet
+              . Governor.localRootPeers)
+              events
+
+          govInProgressPromoteColdSig :: Signal (Set NtNAddr)
+          govInProgressPromoteColdSig =
+            selectDiffusionPeerSelectionState
+              Governor.inProgressPromoteCold
+              events
+
+          govEstablishedPeersSig :: Signal (Set NtNAddr)
+          govEstablishedPeersSig =
+            selectDiffusionPeerSelectionState
+              ( EstablishedPeers.toSet
+              . Governor.establishedPeers)
+              events
+
+          govEstablishedFailuresSig :: Signal (Set NtNAddr)
+          govEstablishedFailuresSig =
+              Signal.keyedLinger
+                180 -- 3 minutes  -- TODO: too eager to reconnect?
+                (fromMaybe Set.empty)
+            . Signal.fromEvents
+            . Signal.selectEvents
+                (\case TracePromoteColdFailed _ _ peer _ _ ->
+                         Just (Set.singleton peer)
+                       --TODO: what about TraceDemoteWarmDone ?
+                       -- these are also not immediate candidates
+                       -- why does the property not fail for not tracking these?
+                       TraceDemoteAsynchronous status
+                         | Set.null failures -> Nothing
+                         | otherwise         -> Just failures
+                         where
+                           failures =
+                             Map.keysSet (Map.filter (==PeerCold) status)
+                       TracePromoteWarmFailed _ _ peer _ ->
+                         Just (Set.singleton peer)
+                       _ -> Nothing
+                )
+            . selectDiffusionPeerSelectionEvents
+            $ events
+
+          -- Left () is a TrJoiningNetwork event
+          -- Right () is a TrKillingNode event
+          trJoinKillSig :: Signal (Either () ())
```

Comment:
Why not use a `Bool` or a even better a:
```
data JoinedOrKilled = Joined | Killed
```

Reply:
Can do :)

Reply:
Done